File: Set 7/World Ladder 2/Ladder2.py
from collections import deque as dq

class node:
    def __init__(self, w, l):
        self.word = w
        self.len = l

def ADJ(a, b):
    count = 0
    for i in range(len(a)):
        if a[i] != b[i]:
            count += 1
    return count == 1

def ladder(beginWord, endWord, wordList):
    q1 = dq([])
    q2 = dq([])
    vis1 = dict()
    v2 = dict()

    start = node(beginWord, 1)
    end = node(endWord, 1)

    vis1[beginWord] = 1
    q1.append(start)
    v2[endWord] = 1
    q2.append(end)

    while q1 and q2:
        curr1 = q1.popleft()
        curr2 = q2.popleft()

        for it in wordList:
            if ADJ(curr1.word, it) and it not in vis1:
                temp = node(it, curr1.len + 1)
                q1.append(temp)
                vis1[it] = curr1.len + 1
                if temp.word == endWord:
                    return temp.len
                if temp.word in v2:
                    return temp.len + v2[temp.word] - 1

        for it in wordList:
            if ADJ(curr2.word, it) and it not in v2:
                temp = node(it, curr2.len + 1)
                q2.append(temp)
                v2[it] = curr2.len + 1
                if temp.word == beginWord:
                    return temp.len
                if temp.word in vis1:
                    return temp.len + vis1[temp.word] - 1

    return 0
t = int(input())
for _ in range(t):
    wordList = list(map(int,input().split()))
    start = str(input())
    end = str(input())
    print(ladder(start, end, wordList))

File: Set 7/World Ladder 2/Ladder2.cpp
#include <iostream>
#include <unordered_set>
#include <queue>
#include <string>
#include <unordered_map>
using namespace std;

class Node {
public:
    string word;
    int len;
    Node(string w, int l) : word(w), len(l) {}
};

bool isAdj(const string &a, const string &b) {
    int count = 0;
    for (int i = 0; i < a.size(); i++) {
        if (a[i] != b[i]) {
            count++;
        }
    }
    return count == 1;
}

int ladderLength(string beginWord, string endWord, unordered_set<string> &wordList) {
    if (wordList.find(endWord) == wordList.end()) return 0;

    queue<Node> q1, q2;
    unordered_map<string, int> vis1, vis2;

    q1.push(Node(beginWord, 1));
    q2.push(Node(endWord, 1));
    vis1[beginWord] = 1;
    vis2[endWord] = 1;

    while (!q1.empty() && !q2.empty()) {
        Node curr1 = q1.front(); q1.pop();
        Node curr2 = q2.front(); q2.pop();

        for (const string &it : wordList) {
            if (isAdj(curr1.word, it) && vis1.find(it) == vis1.end()) {
                Node temp(it, curr1.len + 1);
                q1.push(temp);
                vis1[it] = curr1.len + 1;
                if (temp.word == endWord) return temp.len;
                if (vis2.find(temp.word) != vis2.end()) return temp.len + vis2[temp.word] - 1;
            }
        }

        for (const string &it : wordList) {
            if (isAdj(curr2.word, it) && vis2.find(it) == vis2.end()) {
                Node temp(it, curr2.len + 1);
                q2.push(temp);
                vis2[it] = curr2.len + 1;
                if (temp.word == beginWord) return temp.len;
                if (vis1.find(temp.word) != vis1.end()) return temp.len + vis1[temp.word] - 1;
            }
        }
    }
    return 0;
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        unordered_set<string> wordList;
        string word;
        while (cin >> word) {
            if (word == "-1") break;
            wordList.insert(word);
        }
        string start, end;
        cin >> start >> end;
        cout << ladderLength(start, end, wordList) << endl;
    }
    return 0;
}
